File: apps/azimuth/src/routes/pulsify/tokens.ts
import { createPulsifyDb, dsnTokens } from '@bx-team/stratus/d1';
import { and, eq } from 'drizzle-orm';
import { Hono } from 'hono';
import { z } from 'zod';
import type { Env } from '../../env';
import { ownedProject } from '../../lib/pulsify';
import type { SessionVariables } from '../../middleware/auth';

type PulsifyEnv = { Bindings: Env; Variables: SessionVariables };

const createBodySchema = z.object({
  label: z.string().min(1).max(64).optional(),
});

function generateTokenKey(): string {
  const bytes = crypto.getRandomValues(new Uint8Array(32));
  return [...bytes].map(b => b.toString(16).padStart(2, '0')).join('');
}

export const tokenRoutes = new Hono<PulsifyEnv>();

tokenRoutes.get('/projects/:id/tokens', async c => {
  const session = c.get('authSession');
  if (!session) return c.json({ message: 'Unauthorized' }, 401);
  const db = createPulsifyDb(c.env.PULSIFY_DB);
  const id = c.req.param('id');

  const project = await ownedProject(db, id, session.user.id);
  if (!project) return c.json({ message: 'Project not found' }, 404);

  const tokens = await db
    .select({
      id: dsnTokens.id,
      label: dsnTokens.label,
      revoked: dsnTokens.revoked,
      lastUsedAt: dsnTokens.lastUsedAt,
      createdAt: dsnTokens.createdAt,
    })
    .from(dsnTokens)
    .where(eq(dsnTokens.projectId, id));

  return c.json(tokens);
});

tokenRoutes.post('/projects/:id/tokens', async c => {
  const session = c.get('authSession');
  if (!session) return c.json({ message: 'Unauthorized' }, 401);
  const db = createPulsifyDb(c.env.PULSIFY_DB);
  const id = c.req.param('id');

  const project = await ownedProject(db, id, session.user.id);
  if (!project) return c.json({ message: 'Project not found' }, 404);

  const parsed = createBodySchema.safeParse(await c.req.json().catch(() => ({})));
  if (!parsed.success) return c.json({ message: parsed.error.message }, 400);

  const [token] = await db
    .insert(dsnTokens)
    .values({ projectId: id, key: generateTokenKey(), label: parsed.data.label ?? null })
    .returning({
      id: dsnTokens.id,
      key: dsnTokens.key,
      label: dsnTokens.label,
      createdAt: dsnTokens.createdAt,
    });

  return c.json(token, 201);
});

tokenRoutes.delete('/projects/:id/tokens/:tokenId', async c => {
  const session = c.get('authSession');
  if (!session) return c.json({ message: 'Unauthorized' }, 401);
  const db = createPulsifyDb(c.env.PULSIFY_DB);
  const id = c.req.param('id');
  const tokenId = c.req.param('tokenId');

  const project = await ownedProject(db, id, session.user.id);
  if (!project) return c.json({ message: 'Project not found' }, 404);

  const [token] = await db
    .update(dsnTokens)
    .set({ revoked: true })
    .where(and(eq(dsnTokens.id, tokenId), eq(dsnTokens.projectId, id)))
    .returning({ id: dsnTokens.id });

  if (!token) return c.json({ message: 'Token not found' }, 404);
  return c.body(null, 204);
});

File: apps/azimuth/src/routes/pulsify/tokens.rs
use axum::Json;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use database::models::pulsify as db;
use rand::TryRng;
use util::{ApiError, ApiResult};
use uuid::Uuid;

use crate::auth::project::OwnedProject;
use crate::models::pulsify as api;
use crate::state::AppState;

#[utoipa::path(get, path = "/pulsify/projects/{id}/tokens", tag = "pulsify",
    params(("id" = Uuid, Path, description = "Project id")),
    responses((status = 200, body = Vec<api::Token>), (status = 404)))]
pub async fn list(
    State(state): State<AppState>,
    owned: OwnedProject,
) -> ApiResult<Json<Vec<api::Token>>> {
    let tokens = db::tokens_of(&state.db, owned.project.id).await?;

    Ok(Json(
        tokens
            .into_iter()
            .map(|token| api::Token {
                id: token.id,
                label: token.label,
                revoked: token.revoked,
                last_used_at: token.last_used_at.map(api::iso),
                created_at: api::iso(token.created_at),
            })
            .collect(),
    ))
}

/// Creates a token. This is the only time its key is ever returned.
#[utoipa::path(post, path = "/pulsify/projects/{id}/tokens", tag = "pulsify",
    params(("id" = Uuid, Path, description = "Project id")),
    request_body = api::CreateToken,
    responses((status = 201, body = api::CreatedToken), (status = 404)))]
pub async fn create(
    State(state): State<AppState>,
    owned: OwnedProject,
    Json(body): Json<api::CreateToken>,
) -> ApiResult<(StatusCode, Json<api::CreatedToken>)> {
    let key = random_key()?;
    let label = body
        .label
        .as_deref()
        .map(str::trim)
        .filter(|l| !l.is_empty());

    let token = db::create_token(&state.db, owned.project.id, &key, label).await?;

    Ok((
        StatusCode::CREATED,
        Json(api::CreatedToken {
            id: token.id,
            key,
            label: token.label,
            created_at: api::iso(token.created_at),
        }),
    ))
}

/// Revokes rather than deletes: an ingest token's history has to stay attributable.
#[utoipa::path(delete, path = "/pulsify/projects/{id}/tokens/{token_id}", tag = "pulsify",
    params(
        ("id" = Uuid, Path, description = "Project id"),
        ("token_id" = Uuid, Path, description = "Token id")
    ),
    responses((status = 204), (status = 404)))]
pub async fn revoke(
    State(state): State<AppState>,
    owned: OwnedProject,
    Path(path): Path<(Uuid, Uuid)>,
) -> ApiResult<StatusCode> {
    let (_, token_id) = path;
    if !db::revoke_token(&state.db, owned.project.id, token_id).await? {
        return Err(ApiError::NotFound("Token not found".into()));
    }

    Ok(StatusCode::NO_CONTENT)
}

fn random_key() -> ApiResult<String> {
    let mut bytes = [0u8; 32];
    rand::rngs::SysRng
        .try_fill_bytes(&mut bytes)
        .map_err(|error| ApiError::internal(error.to_string()))?;
    Ok(hex::encode(bytes))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn keys_are_unpredictable_and_the_expected_width() {
        let first = random_key().unwrap();
        assert_eq!(first.len(), 64);
        assert!(first.bytes().all(|byte| byte.is_ascii_hexdigit()));
        assert_ne!(first, random_key().unwrap());
    }
}
